crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -66,10 +65,21 @@ impl FormatRule<Expr, PyFormatContext<'_>> for FormatExpr {
     fn fmt(&self, expression: &Expr, f: &mut PyFormatter) -> FormatResult<()> {
         let parentheses = self.parentheses;
 
+        // All expressions that can have parentheses can be in fluent style. We need to exclude top
+        // expressions since they don't get parenthesized
+        let call_chain_layout = if matches!(
+            f.context().node_level(),
+            NodeLevel::CompoundStatement | NodeLevel::TopLevel
+        ) {
+            CallChainLayout::Default
+        } else {
+            CallChainLayout::from_expression(expression, f.context().source())
+        };
```

Comment:
I played around with your PR locally, and moving the call chain formatting out of the `Expr` seems possible. It has the added benefit that we avoid adding unnecessary groups around `Subsccript` and `Attribute` accessors

```patch
Subject: [PATCH] Move Fluent style out of `Expr` formatting
---
Index: crates/ruff_python_formatter/src/expression/expr_subscript.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/ruff_python_formatter/src/expression/expr_subscript.rs b/crates/ruff_python_formatter/src/expression/expr_subscript.rs
--- a/crates/ruff_python_formatter/src/expression/expr_subscript.rs	(revision 54b78ac7bdf0728ea0f0d91a426e460046d827e6)
+++ b/crates/ruff_python_formatter/src/expression/expr_subscript.rs	(date 1691133618832)
@@ -34,6 +34,17 @@
             ctx: _,
         } = item;
 
+        let call_chain_layout = match self.call_chain_layout {
+            CallChainLayout::Default => {
+                if f.context().node_level().is_parenthesized() {
+                    CallChainLayout::from_expression(AnyNodeRef::from(item), f.context().source())
+                } else {
+                    CallChainLayout::NonFluent
+                }
+            }
+            layout @ (CallChainLayout::Fluent | CallChainLayout::NonFluent) => layout,
+        };
+
         let comments = f.context().comments().clone();
         let dangling_comments = comments.dangling_comments(item.as_any_node_ref());
         debug_assert!(
@@ -42,9 +53,9 @@
         );
 
         let format_value = format_with(|f| match value.as_ref() {
-            Expr::Attribute(expr) => expr.format().with_options(self.call_chain_layout).fmt(f),
-            Expr::Call(expr) => expr.format().with_options(self.call_chain_layout).fmt(f),
-            Expr::Subscript(expr) => expr.format().with_options(self.call_chain_layout).fmt(f),
+            Expr::Attribute(expr) => expr.format().with_options(call_chain_layout).fmt(f),
+            Expr::Call(expr) => expr.format().with_options(call_chain_layout).fmt(f),
+            Expr::Subscript(expr) => expr.format().with_options(call_chain_layout).fmt(f),
             _ => value.format().fmt(f),
         });
 
@@ -91,10 +102,16 @@
     fn needs_parentheses(
         &self,
         _parent: AnyNodeRef,
-        _context: &PyFormatContext,
+        context: &PyFormatContext,
     ) -> OptionalParentheses {
         {
-            OptionalParentheses::Never
+            if CallChainLayout::from_expression(self.into(), context.source())
+                == CallChainLayout::Fluent
+            {
+                OptionalParentheses::Multiline
+            } else {
+                self.value.needs_parentheses(self.into(), context)
+            }
         }
     }
 }
Index: crates/ruff_python_formatter/src/expression/expr_call.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/ruff_python_formatter/src/expression/expr_call.rs b/crates/ruff_python_formatter/src/expression/expr_call.rs
--- a/crates/ruff_python_formatter/src/expression/expr_call.rs	(revision 54b78ac7bdf0728ea0f0d91a426e460046d827e6)
+++ b/crates/ruff_python_formatter/src/expression/expr_call.rs	(date 1691134353539)
@@ -1,6 +1,7 @@
 use crate::expression::CallChainLayout;
 use ruff_formatter::{write, FormatRuleWithOptions};
 use ruff_python_ast::node::AnyNodeRef;
+use ruff_python_ast::Expr::Call;
 use ruff_python_ast::{Expr, ExprCall};
 
 use crate::expression::parentheses::{NeedsParentheses, OptionalParentheses};
@@ -29,14 +30,41 @@
             arguments,
         } = item;
 
-        match func.as_ref() {
-            Expr::Attribute(expr) => expr.format().with_options(self.call_chain_layout).fmt(f)?,
-            Expr::Call(expr) => expr.format().with_options(self.call_chain_layout).fmt(f)?,
-            Expr::Subscript(expr) => expr.format().with_options(self.call_chain_layout).fmt(f)?,
-            _ => func.format().fmt(f)?,
-        }
+        let call_chain_layout = match self.call_chain_layout {
+            CallChainLayout::Default => {
+                if f.context().node_level().is_parenthesized() {
+                    CallChainLayout::from_expression(AnyNodeRef::from(item), f.context().source())
+                } else {
+                    CallChainLayout::NonFluent
+                }
+            }
+            layout @ (CallChainLayout::Fluent | CallChainLayout::NonFluent) => layout,
+        };
+
+        let fmt_inner = format_with(|f| {
+            match func.as_ref() {
+                Expr::Attribute(expr) => expr.format().with_options(call_chain_layout).fmt(f)?,
+                Expr::Call(expr) => expr.format().with_options(call_chain_layout).fmt(f)?,
+                Expr::Subscript(expr) => expr.format().with_options(call_chain_layout).fmt(f)?,
+                _ => func.format().fmt(f)?,
+            }
 
-        write!(f, [arguments.format()])
+            arguments.format().fmt(f)
+        });
+
+        // Allow to indent the parentheses while
+        // ```python
+        // g1 = (
+        //     queryset.distinct().order_by(field.name).values_list(field_name_flat_long_long=True)
+        // )
+        // ```
+        if call_chain_layout == CallChainLayout::Fluent
+            && self.call_chain_layout == CallChainLayout::Default
+        {
+            group(&fmt_inner).fmt(f)
+        } else {
+            fmt_inner.fmt(f)
+        }
     }
 }
 
@@ -46,6 +74,12 @@
         _parent: AnyNodeRef,
         context: &PyFormatContext,
     ) -> OptionalParentheses {
-        self.func.needs_parentheses(self.into(), context)
+        if CallChainLayout::from_expression(self.into(), context.source())
+            == CallChainLayout::Fluent
+        {
+            OptionalParentheses::Multiline
+        } else {
+            self.func.needs_parentheses(self.into(), context)
+        }
     }
 }
Index: crates/ruff_python_formatter/src/expression/mod.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/ruff_python_formatter/src/expression/mod.rs b/crates/ruff_python_formatter/src/expression/mod.rs
--- a/crates/ruff_python_formatter/src/expression/mod.rs	(revision 54b78ac7bdf0728ea0f0d91a426e460046d827e6)
+++ b/crates/ruff_python_formatter/src/expression/mod.rs	(date 1691134350626)
@@ -65,17 +65,6 @@
     fn fmt(&self, expression: &Expr, f: &mut PyFormatter) -> FormatResult<()> {
         let parentheses = self.parentheses;
 
-        // All expressions that can have parentheses can be in fluent style. We need to exclude top
-        // expressions since they don't get parenthesized
-        let call_chain_layout = if matches!(
-            f.context().node_level(),
-            NodeLevel::CompoundStatement | NodeLevel::TopLevel
-        ) {
-            CallChainLayout::Default
-        } else {
-            CallChainLayout::from_expression(expression, f.context().source())
-        };
-
         let format_expr = format_with(|f| match expression {
             Expr::BoolOp(expr) => expr.format().with_options(Some(parentheses)).fmt(f),
             Expr::NamedExpr(expr) => expr.format().fmt(f),
@@ -93,12 +82,12 @@
             Expr::Yield(expr) => expr.format().fmt(f),
             Expr::YieldFrom(expr) => expr.format().fmt(f),
             Expr::Compare(expr) => expr.format().with_options(Some(parentheses)).fmt(f),
-            Expr::Call(expr) => expr.format().with_options(call_chain_layout).fmt(f),
+            Expr::Call(expr) => expr.format().fmt(f),
             Expr::FormattedValue(expr) => expr.format().fmt(f),
             Expr::JoinedStr(expr) => expr.format().fmt(f),
             Expr::Constant(expr) => expr.format().fmt(f),
-            Expr::Attribute(expr) => expr.format().with_options(call_chain_layout).fmt(f),
-            Expr::Subscript(expr) => expr.format().with_options(call_chain_layout).fmt(f),
+            Expr::Attribute(expr) => expr.format().fmt(f),
+            Expr::Subscript(expr) => expr.format().fmt(f),
             Expr::Starred(expr) => expr.format().fmt(f),
             Expr::Name(expr) => expr.format().fmt(f),
             Expr::List(expr) => expr.format().fmt(f),
@@ -128,17 +117,7 @@
 
             let mut f = WithNodeLevel::new(level, f);
 
-            // Allow to indent the parentheses while
-            // ```python
-            // g1 = (
-            //     queryset.distinct().order_by(field.name).values_list(field_name_flat_long_long=True)
-            // )
-            // ```
-            if call_chain_layout == CallChainLayout::Fluent {
-                write!(f, [group(&format_expr)])
-            } else {
-                write!(f, [format_expr])
-            }
+            write!(f, [format_expr])
         }
     }
 }
@@ -175,35 +154,6 @@
             parenthesize,
         } = self;
 
-        // Fluent style means that when we break we always add parentheses, so we don't need the
-        // checks below for existing parentheses, comments or `needs_parentheses`
-        if CallChainLayout::from_expression(expression, f.context().source())
-            == CallChainLayout::Fluent
-        {
-            match expression {
-                Expr::Attribute(expr) => {
-                    return parenthesize_if_expands(&group(
-                        &expr.format().with_options(CallChainLayout::Fluent),
-                    ))
-                    .fmt(f)
-                }
-                Expr::Call(expr) => {
-                    return parenthesize_if_expands(&group(
-                        &expr.format().with_options(CallChainLayout::Fluent),
-                    ))
-                    .fmt(f)
-                }
-                Expr::Subscript(expr) => {
-                    return parenthesize_if_expands(&group(
-                        &expr.format().with_options(CallChainLayout::Fluent),
-                    ))
-                    .fmt(f)
-                }
-                // `call_chain_layout` checks we're in one of the above
-                _ => unreachable!(),
-            };
-        }
-
         let comments = f.context().comments();
         let preserve_parentheses = parenthesize.is_optional()
             && is_expression_parenthesized((*expression).into(), f.context().source());
@@ -527,30 +477,36 @@
 /// ```
 #[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
 pub enum CallChainLayout {
+    /// The root of a call chain
     #[default]
     Default,
+
+    /// A nested call chain element that uses fluent style.
     Fluent,
+
+    /// A nested call chain element not using fluent style.
+    NonFluent,
 }
 
 impl CallChainLayout {
-    pub(crate) fn from_expression(mut expr: &Expr, source: &str) -> Self {
+    pub(crate) fn from_expression(mut expr: AnyNodeRef, source: &str) -> Self {
         let mut attributes_after_parentheses = 0;
         loop {
             match expr {
-                Expr::Attribute(ast::ExprAttribute { value, .. }) => {
+                AnyNodeRef::ExprAttribute(ast::ExprAttribute { value, .. }) => {
                     // `f().x` | `data[:100].T`
                     if matches!(value.as_ref(), Expr::Call(_) | Expr::Subscript(_)) {
                         attributes_after_parentheses += 1;
                     }
-                    expr = value;
+                    expr = AnyNodeRef::from(value.as_ref());
                 }
-                Expr::Call(ast::ExprCall { func: inner, .. })
-                | Expr::Subscript(ast::ExprSubscript { value: inner, .. }) => {
-                    expr = inner;
+                AnyNodeRef::ExprCall(ast::ExprCall { func: inner, .. })
+                | AnyNodeRef::ExprSubscript(ast::ExprSubscript { value: inner, .. }) => {
+                    expr = AnyNodeRef::from(inner.as_ref());
                 }
                 _ => {
                     // We to format the following in fluent style: `f2 = (a).w().t(1,)`
-                    if is_expression_parenthesized(AnyNodeRef::from(expr), source) {
+                    if is_expression_parenthesized(expr, source) {
                         attributes_after_parentheses += 1;
                     }
                     break;
@@ -558,7 +514,7 @@
             }
         }
         if attributes_after_parentheses < 2 {
-            CallChainLayout::Default
+            CallChainLayout::NonFluent
         } else {
             CallChainLayout::Fluent
         }
Index: crates/ruff_python_formatter/src/expression/expr_attribute.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/ruff_python_formatter/src/expression/expr_attribute.rs b/crates/ruff_python_formatter/src/expression/expr_attribute.rs
--- a/crates/ruff_python_formatter/src/expression/expr_attribute.rs	(revision 54b78ac7bdf0728ea0f0d91a426e460046d827e6)
+++ b/crates/ruff_python_formatter/src/expression/expr_attribute.rs	(date 1691133618832)
@@ -31,6 +31,17 @@
             ctx: _,
         } = item;
 
+        let call_chain_layout = match self.call_chain_layout {
+            CallChainLayout::Default => {
+                if f.context().node_level().is_parenthesized() {
+                    CallChainLayout::from_expression(AnyNodeRef::from(item), f.context().source())
+                } else {
+                    CallChainLayout::NonFluent
+                }
+            }
+            layout @ (CallChainLayout::Fluent | CallChainLayout::NonFluent) => layout,
+        };
+
         let needs_parentheses = matches!(
             value.as_ref(),
             Expr::Constant(ExprConstant {
@@ -51,18 +62,18 @@
         } else {
             match value.as_ref() {
                 Expr::Attribute(expr) => {
-                    expr.format().with_options(self.call_chain_layout).fmt(f)?;
+                    expr.format().with_options(call_chain_layout).fmt(f)?;
                 }
                 Expr::Call(expr) => {
-                    expr.format().with_options(self.call_chain_layout).fmt(f)?;
-                    if self.call_chain_layout == CallChainLayout::Fluent {
+                    expr.format().with_options(call_chain_layout).fmt(f)?;
+                    if call_chain_layout == CallChainLayout::Fluent {
                         // Format the dot on its own line
                         soft_line_break().fmt(f)?;
                     }
                 }
                 Expr::Subscript(expr) => {
-                    expr.format().with_options(self.call_chain_layout).fmt(f)?;
-                    if self.call_chain_layout == CallChainLayout::Fluent {
+                    expr.format().with_options(call_chain_layout).fmt(f)?;
+                    if call_chain_layout == CallChainLayout::Fluent {
                         // Format the dot on its own line
                         soft_line_break().fmt(f)?;
                     }
@@ -75,7 +86,7 @@
             hard_line_break().fmt(f)?;
         }
 
-        if self.call_chain_layout == CallChainLayout::Fluent {
+        if call_chain_layout == CallChainLayout::Fluent {
             // Fluent style has line breaks before the dot
             // ```python
             // blogs3 = (
@@ -139,7 +150,11 @@
         context: &PyFormatContext,
     ) -> OptionalParentheses {
         // Checks if there are any own line comments in an attribute chain (a.b.c).
-        if context
+        if CallChainLayout::from_expression(self.into(), context.source())
+            == CallChainLayout::Fluent
+        {
+            OptionalParentheses::Multiline
+        } else if context
             .comments()
             .dangling_comments(self)
             .iter()

```

There's room to remove some of the duplication in `Subscript`, `Attribute`, and the call formatting. I also haven't tested if there are no regressions on the django projects. 

I would prefer this solution because it avoids by-passing `NeedsParentheses`, `can_omit_parentheses`, and the changes are local to the affected expression types. But interested to hear your opinion

Reply:
Should have checked first, there's a case of unstable formatting now:
```python
class s():s=((()).r(Q(e="efinitive Guide to Django: Web Development Done Right")())).order_by()
```